service/domain/blobs/replication/has_handler.go

```diff
@@ -0,0 +1,91 @@
+package replication
+
+import (
+	"context"
+
+	"github.com/boreq/errors"
+	"github.com/planetary-social/go-ssb/logging"
+	"github.com/planetary-social/go-ssb/service/domain/blobs"
+	"github.com/planetary-social/go-ssb/service/domain/refs"
+	"github.com/planetary-social/go-ssb/service/domain/transport"
+)
+
+type BlobStorage interface {
+	Has(id refs.Blob) (bool, error)
+}
+
+type WantListRepository interface {
+	WantListContains(id refs.Blob) (bool, error)
+	DeleteFromWantList(id refs.Blob) error
+}
+
+type Downloader interface {
+	Download(ctx context.Context, peer transport.Peer, blob refs.Blob) error
+}
+
+type HasHandler struct {
+	storage    BlobStorage
+	wantList   WantListRepository
+	downloader Downloader
+	logger     logging.Logger
+}
+
+func NewHasHandler(
+	storage BlobStorage,
+	wantList WantListRepository,
+	downloader Downloader,
+	logger logging.Logger,
+) *HasHandler {
+	return &HasHandler{
+		storage:    storage,
+		wantList:   wantList,
+		downloader: downloader,
+		logger:     logger.New("has_handler"),
+	}
+}
+
+func (d *HasHandler) OnHasReceived(ctx context.Context, peer transport.Peer, blob refs.Blob, size blobs.Size) {
+	d.logger.WithField("blob", blob.String()).Debug("has")
+
+	if err := d.onHasReceived(ctx, peer, blob, size); err != nil {
+		d.logger.WithError(err).Error("failed to download a blob")
+	}
+}
+
+func (d *HasHandler) onHasReceived(ctx context.Context, peer transport.Peer, blob refs.Blob, size blobs.Size) error {
```

Comment:
Opinion on moving `HasHandler` to application and making it a command? This would give us normal access to transactions for example. This is starting to look like application code to me especially because I want to emit events after the blobs were downloaded at some point.

Reply:
@czeslavo

Reply:
Yeah, I agree it looks like a good candidate to become an app command.